Sources/OpenAPIAsyncHTTPClient/AsyncHTTPClientTransport.swift

```diff
@@ -150,57 +151,94 @@ public struct AsyncHTTPClientTransport: ClientTransport {
             throw Error.invalidRequestURL(request: request, baseURL: baseURL)
         }
-        baseUrlComponents.percentEncodedPath += request.path
-        baseUrlComponents.percentEncodedQuery = request.query
+        baseUrlComponents.percentEncodedPath += request.soar_pathOnly
+        baseUrlComponents.percentEncodedQuery = request.soar_query.map(String.init)
         guard let url = baseUrlComponents.url else {
             throw Error.invalidRequestURL(request: request, baseURL: baseURL)
         }
         var clientRequest = HTTPClientRequest(url: url.absoluteString)
         clientRequest.method = request.method.asHTTPMethod
         for header in request.headerFields {
-            clientRequest.headers.add(name: header.name.lowercased(), value: header.value)
+            clientRequest.headers.add(name: header.name.canonicalName, value: header.value)
         }
-        if let body = request.body {
-            clientRequest.body = .bytes(body)
+        if let body {
+            let length: HTTPClientRequest.Body.Length
+            switch body.length {
+            case .unknown:
+                length = .unknown
+            case .known(let count):
+                length = .known(count)
+            }
+            clientRequest.body = .stream(
+                body.map { .init(bytes: $0) },
+                length: length
+            )
         }
         return clientRequest
     }
 
     /// Converts the received URLResponse into the shared Response.
     internal static func convertResponse(
-        _ httpResponse: HTTPClientResponse
-    ) async throws -> OpenAPIRuntime.Response {
-        let headerFields: [OpenAPIRuntime.HeaderField] = httpResponse
-            .headers
-            .map { .init(name: $0, value: $1) }
-        let body = try await httpResponse.body.collect(upTo: .max)
-        let bodyData = Data(buffer: body, byteTransferStrategy: .noCopy)
-        let response = OpenAPIRuntime.Response(
-            statusCode: Int(httpResponse.status.code),
-            headerFields: headerFields,
-            body: bodyData
+        method: HTTPRequest.Method,
+        httpResponse: HTTPClientResponse
+    ) async throws -> (HTTPResponse, HTTPBody?) {
+
+        var headerFields: HTTPFields = [:]
+        for header in httpResponse.headers {
+            headerFields[.init(header.name)!] = header.value
+        }
+
+        let length: HTTPBody.Length
+        if let lengthHeaderString = headerFields[.contentLength],
+            let lengthHeader = Int(lengthHeaderString)
+        {
+            length = .known(lengthHeader)
+        } else {
+            length = .unknown
+        }
+
+        let body: HTTPBody?
+        switch method {
+        case .head, .connect, .trace:
+            body = nil
+        default:
+            body = HTTPBody(
+                httpResponse.body.map { $0.readableBytesView },
+                length: length,
+                iterationBehavior: .single
+            )
+        }
```

Comment:
That's the API contract with the runtime library. For HTTP methods that are not supposed to return bodies, we return nil, instead of an empty body. Some previous discussion on this can be found here: https://forums.swift.org/t/proposal-soar-0005-adopting-the-swift-http-types-package/67185/5

Reply:
Oh, I missed that change. So yes, the HTTP wire protocol differentiates between no body and an empty body. But there is no semantic difference. A HTTP GET request is only ever allowed to have no request body send at all over the wire and a HTTP GET request with a request body is not a valid request, even if it is empty. This difference is only important to the implementations of AHC and URLSession but isn't useful to the user at all. I don't see semantic value in an optional request or response body as it is semantically equivalent to an empty body and the HTTPClient libraries should just take care of the wire protocol details.